x-pack/plugins/alerting/server/task_runner/action_scheduler/action_scheduler.ts

```diff
@@ -0,0 +1,605 @@
+import {
+  AlertInstanceContext,
+  AlertInstanceState,
+  RuleAction,
+  RuleTypeParams,
+  RuleTypeState,
+  SanitizedRule,
+  RuleAlertData,
+  RuleSystemAction,
+} from '../../../common';
+import {
+  generateActionHash,
+  getSummaryActionsFromTaskState,
+  getSummaryActionTimeBounds,
+  isActionOnInterval,
+} from './rule_action_helper';
+import { RULE_SAVED_OBJECT_TYPE } from '../../saved_objects';
+import { ConnectorAdapter } from '../../connector_adapters/types';
+import { withAlertingSpan } from '../lib';
+import * as schedulers from './schedulers';
+
+interface LogAction {
+  id: string;
+  typeId: string;
+  alertId?: string;
+  alertGroup?: string;
+  alertSummary?: {
+    new: number;
+    ongoing: number;
+    recovered: number;
+  };
+}
+
+interface RunSummarizedActionArgs {
+  action: RuleAction;
+  summarizedAlerts: CombinedSummarizedAlerts;
+  spaceId: string;
+  bulkActions: EnqueueExecutionOptions[];
+}
+
+interface RunSystemActionArgs<Params extends RuleTypeParams> {
+  action: RuleSystemAction;
+  connectorAdapter: ConnectorAdapter;
+  summarizedAlerts: CombinedSummarizedAlerts;
+  rule: SanitizedRule<Params>;
+  ruleProducer: string;
+  spaceId: string;
+  bulkActions: EnqueueExecutionOptions[];
+}
+
+interface RunActionArgs<
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+> {
+  action: RuleAction;
+  alert: Alert<State, Context, ActionGroupIds | RecoveryActionGroupId>;
+  ruleId: string;
+  spaceId: string;
+  bulkActions: EnqueueExecutionOptions[];
+}
+
+export interface RunResult {
+  throttledSummaryActions: ThrottledActions;
+}
+
+export class ActionScheduler<
+  Params extends RuleTypeParams,
+  ExtractedParams extends RuleTypeParams,
+  RuleState extends RuleTypeState,
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string,
+  AlertData extends RuleAlertData
+> {
+  private readonly schedulers: Array<
+    IActionScheduler<State, Context, ActionGroupIds, RecoveryActionGroupId>
+  > = [];
+
+  private ephemeralActionsToSchedule: number;
+  private CHUNK_SIZE = 1000;
+  private ruleTypeActionGroups?: Map<ActionGroupIds | RecoveryActionGroupId, string>;
+  private previousStartedAt: Date | null;
+
+  constructor(
+    private readonly context: ActionSchedulerOptions<
+      Params,
+      ExtractedParams,
+      RuleState,
+      State,
+      Context,
+      ActionGroupIds,
+      RecoveryActionGroupId,
+      AlertData
+    >
+  ) {
+    this.ephemeralActionsToSchedule = context.taskRunnerContext.maxEphemeralActionsPerRule;
+    this.ruleTypeActionGroups = new Map(
+      context.ruleType.actionGroups.map((actionGroup) => [actionGroup.id, actionGroup.name])
+    );
+    this.previousStartedAt = context.previousStartedAt;
+
+    for (const [_, scheduler] of Object.entries(schedulers)) {
```

Comment:
This grabs every scheduler from the `schedulers/index.ts` file

Reply:
I tried moving `schedulers` into a static variable

```
static schedulers: Array<
    IActionScheduler<State, Context, ActionGroupIds, RecoveryActionGroupId>
  > = [];
```

 but I get a TS warning: `Static members cannot reference class type parameters`